### ISSUE TITLE
Ensure that Python and Cython versions of abel.direct produce identical results

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#58]
As mentioned in PR https://github.com/PyAbel/PyAbel/pull/52 and https://github.com/PyAbel/PyAbel/pull/52#issuecomment-164826567, the results of the Python and Cython implementations of the direct fabel/iabel transforms are similar, but not identical. At some point, we should track down the cause of this and figure out how to make them exactly the same.

**Comment by rth** [comment]
At this points, it appears clear that I won't have the bandwidth to finish this, so if anyone is motivated to take up this issue they can feel free to do so.

**Comment by DanHickstein** [comment]
It seems like the simpson's rule integration might be leading us astray. Perhaps if we switch both methods to a `trapz` integration, then the issue will be easier to diagnose. @rth, if you have a moment, can you demonstrate how to switch the Cython code to a trapz integration?

**Comment by stggh** [comment]
With the `O2-ANU1024.txt` image `direct-C` is similar to `direct-python`. This is a plot of `row=512` of the transformed image:
![o2-directpc](https://user-images.githubusercontent.com/10932229/34464890-e9844fd6-eee8-11e7-8147-ffc6585e234b.png)
```python
im = np.loadtxt('O2-ANU1024.txt.bz2')

directC = abel.Transform(im, method='direct', center='convolution',
                         transform_options=dict(backend='c', dr=1)).transform

directP = abel.Transform(im, method='direct', center='convolution',
                         transform_options=dict(backend='python', dr=1)).transform
```

However, for the *1D* analytical functions there is a substantial difference, which may point to some truncation error in the true python version:
![ck](https://user-images.githubusercontent.com/10932229/34464892-0269e7cc-eee9-11e7-9967-9ded112ae6b7.png)

Interestingly, `direct-c` mirrors `hansenlaw`. As usual `three_point` is the most accurate.

Transform code:
```python
import numpy as np
import abel
import matplotlib.pyplot as plt

n = 101
f = abel.tools.analytical.TransformPair(n, profile=8)

hlf = abel.hansenlaw.hansenlaw_transform(f.func, dr=f.dr, direction='forward')
hli = abel.hansenlaw.hansenlaw_transform(f.abel, dr=f.dr)

tpi = abel.dasch.three_point_transform(f.abel, dr=f.dr, basis_dir=None)
dic = abel.direct.direct_transform(f.abel, dr=f.dr, backend='c')
dip = abel.direct.direct_transform(f.abel, dr=f.dr, backend='python')
```

**Comment by DanHickstein** [comment]
Okay, I think that I've nearly cracked this puzzle.

The basic mistake we were making is not calling `scipy.integrate.simps` on the actual array that we want to integrate, but instead setting the values where `y<=r` to zero. This is exactly the same thing if we are using `np.sum`, but is we are using `np.trapz` or `scipy.integrate.simps`, then we basically create an off-by-one error at the edge. This behavior is illustrated by comparing the results of the "one-loop" transform with the "two-loop" transform where `simps` is used correctly: https://gist.github.com/DanHickstein/d2f2b66de5cebddc4b7ee144a229801c

I'm not sure if that the current Cython code in `direct.pyx` actually takes a correct Simpson's rule integration. I switched everything to trapezoidal integration (`np.trapz` for the Direct-P version), and the results are almost identical to the previous direct-C implementation.

For the direct-P version, I needed to include a correction:
```python
out[i, :] = out[i, :] - 0.5*int_func(P*mask2, axis=1, **int_opts)
```
which fixes the issue with using zeros to build our masked array.

Since Simpson's rule doesn't play well with odd-sized arrays (which is now our default), and since a Trapezoidal integration seems to provide near-identical functionality, I switched all of the defaults to Trapezoidal integration rather than Simpson's Rule. @rth, do you have any objection to this?

I started a PR where the changes can be seen:
https://github.com/PyAbel/PyAbel/pull/204

Here is a sample comparison between the new Direct-C and Direct-P:

![figure_1](https://user-images.githubusercontent.com/1107796/34471170-dc16757a-eefe-11e7-90e2-0f6edf3ae9d6.png)

The code used to make the plot is here: https://gist.github.com/DanHickstein/97c8a69d94525e318edba2eb6bd1f577

**Comment by DanHickstein** [comment]
This was addressed by #204. The Direct-C and Direct-Python methods now produce pretty darn identical results.
